datafusion/functions/src/unicode/initcap.rs

```diff
@@ -149,13 +162,16 @@ fn initcap_string(input: Option<&str>) -> Option<String> {
         let mut prev_is_alphanumeric = false;
 
         for c in s.chars() {
-            let transformed = if prev_is_alphanumeric {
-                c.to_ascii_lowercase()
+            if prev_is_alphanumeric {
+                for lc in c.to_lowercase() {
+                    result.push(lc);
+                }
             } else {
-                c.to_ascii_uppercase()
-            };
-            result.push(transformed);
-            prev_is_alphanumeric = c.is_ascii_alphanumeric();
+                for uc in c.to_uppercase() {
+                    result.push(uc);
+                }
+            }
```

Comment:
Does this change slow down the ASCII-only case? If yes, maybe we can consider keeping a faster path for ASCII-only cases. (like https://github.com/apache/datafusion/issues/12306)

Reply:
@goldmedal thanks for reviewing

> Does this change slow down the ASCII-only case? If yes, maybe we can consider keeping a faster path for ASCII-only cases. (like https://github.com/apache/datafusion/issues/12306)

Seems like keeping a fast path for ASCII-only cases is beneficial. I will check and update it soon.